consumer/consumererror/README.md

```diff
@@ -0,0 +1,231 @@
+package provide functionality for communicating details about
+the error for use upstream in the pipeline. Ideally the error returned by a
+component in its `consume` function should be from this package.
+
+## Error classes
+
+**Retryable**: Errors are retryable if re-submitting data to a sink may result
+in a successful submission.
+
+**Permanent**: Errors are permanent if submission will always fail for the
+current data. Errors are considered permanent unless they are explicitly marked
+as retryable.
+
+## Use cases
+
+**Retry logic**: Errors should be allowed to include information necessary to
+perform retries.
+
+**Indicating partial success**: Errors can indicate that not all items were
+accepted, for example as in an OTLP partial success message.
+
+**Communicating network error codes**: Errors should allow embedding information
+necessary for the Collector to act as a proxy for a backend, i.e. relay a status
+code returned from a backend in a response to a system upstream from the
+Collector.
+
+## Current targets for using errors
+
+**Receivers**: Receivers should be able to consume multiple errors downstream
+and determine the best course of action based on the user's configuration. This
+may entail either keeping the retry queue inside of the Collector by having the
+receiver keep track of retries, or may involve having the caller manage the
+retry queue by returning a retryable network code with relevant information.
+
+**exporterhelper**: When an exporter returns a retryable error, the
+exporterhelper can use this information to manage the sending queue if it is
+enabled. Permanent errors will be forwarded back up the pipeline.
+
+## Creating Errors
+
+Errors can be created by calling `consumererror.New(err, opts...)` where `err`
+is the underlying error, and `opts` is one of the provided options for supplying
+additional metadata:
+
+- `consumererror.WithRetry`
+- `consumererror.WithPartial`
+- `consumererror.WithGRPCStatus`
+- `consumererror.WithHTTPStatus`
+- `consumererror.WithMetadata`
+
+**Example**:
+
+```go
+consumererror.New(err,
+  consumererror.WithRetry(
+    componentID,
+    consumerrerror.WithRetryDelay(10 * time.Second)
+  ),
+  consumererror.WithGRPCStatus(codes.InvalidArgument),
+)
+```
+
+### Retrying data submission
+
+*NOTE*: This section is a draft and will be developed separately from the rest of this proposal.
+
+If an error is transient, the `WithRetry` option corresponding to the relevant
+signal should be used to indicate that the error is retryable and to pass on any
+retry settings. These settings can come from the data sink or be determined by
+the component, such as through runtime conditions or from user settings.
+
+The data for the retry will be provided by the component performing the retry.
+
+**Note**: If retry information is not included in an error, the error will be
+considered permanent and will not be retried.
+
+**Usage:**
+
+```go
+consumererror.WithRetry(
+  componentID,
```

Comment:
What do I do with the "component" if I know that a specific component failed? Especially that I don't know the structure of the pipelines.

Reply:
That's actually a good point, we should have some other pipeline component add the component ID automatically. The goal is that we will be including the failed pipeline branch inside this error so only retryable branches are retried.

Note that retryable errors are outside the scope of what will be implemented in this PR, they're a lot more complicated than e.g. HTTP/gRPC codes.